File: scripts/prepare_sheets_import.py
import os
import glob
import csv
import json
import re

DOWNLOADS_DIR = r"C:\Users\ETIVE\Downloads"
OUTPUT_DIR = os.path.join(os.path.dirname(__file__), "..", "data")
os.makedirs(OUTPUT_DIR, exist_ok=True)

def normalize(text):
    if not text:
        return ""
    return re.sub(r"\s+", " ", str(text).strip())

def parse_csv(filepath):
    rows = []
    with open(filepath, "r", encoding="utf-8-sig", errors="replace") as f:
        reader = csv.reader(f)
        header = next(reader, None)
        if not header:
            return rows
        
        headers = [h.strip().lower() for h in header]
        sr_idx = next((i for i, h in enumerate(headers) if h in ["sr.", "sr", "serial", "id"]), 0)
        sub_idx = next((i for i, h in enumerate(headers) if "subject" in h), 1)
        top_idx = next((i for i, h in enumerate(headers) if "topic" in h and "sub" not in h), 2)
        subtop_idx = next((i for i, h in enumerate(headers) if "subtopic" in h or "sub_topic" in h), 3)

        for line in reader:
            if not line:
                continue
            sr = line[sr_idx] if sr_idx < len(line) else ""
            subj = normalize(line[sub_idx] if sub_idx < len(line) else "")
            top = normalize(line[top_idx] if top_idx < len(line) else "")
            subtop = normalize(line[subtop_idx] if subtop_idx < len(line) else "")

            if subj and top and subtop:
                rows.append({"sr": sr, "subject": subj, "topic": top, "subtopic": subtop})
    return rows

def parse_md_table(filepath, default_subject):
    rows = []
    with open(filepath, "r", encoding="utf-8", errors="replace") as f:
        for line in f:
            line = line.strip()
            if not line.startswith("|") or "---" in line or "Sr." in line:
                continue
            cols = [normalize(c) for c in line.split("|") if c.strip()]
            if len(cols) >= 3:
                sr = cols[0]
                if len(cols) >= 4:
                    subj = cols[1] or default_subject
                    top = cols[2]
                    subtop = cols[3]
                else:
                    subj = default_subject
                    top = cols[1]
                    subtop = cols[2]
                if top and subtop:
                    rows.append({"sr": sr, "subject": subj, "topic": top, "subtopic": subtop})
    return rows

def main():
    print(f"Scanning taxonomy files in: {DOWNLOADS_DIR}")
    files = glob.glob(os.path.join(DOWNLOADS_DIR, "*taxonomy*.csv")) + glob.glob(os.path.join(DOWNLOADS_DIR, "*taxonomy*.md"))
    
    all_rows = []
    seen_keys = set()
    duplicates_count = 0

    for f in sorted(files):
        fname = os.path.basename(f)
        if f.endswith(".csv"):
            parsed = parse_csv(f)
        elif f.endswith(".md"):
            subj_guess = "Physics" if "physics" in fname.lower() else "Economics" if "economics" in fname.lower() else "General Knowledge"
            parsed = parse_md_table(f, subj_guess)
        else:
            continue

        added = 0
        for r in parsed:
            key = f"{r['subject'].lower()}|{r['topic'].lower()}|{r['subtopic'].lower()}"
            if key in seen_keys:
                duplicates_count += 1
                continue
            seen_keys.add(key)
            all_rows.append(r)
            added += 1
        
        print(f"  * {fname:<60} : {len(parsed):>5} found -> {added:>5} unique added")

    # Sequential re-indexing of source_sr
    for idx, r in enumerate(all_rows):
        r["sr"] = idx + 1

    print("=" * 80)
    print(f"TOTAL UNIQUE TAXONOMY ROWS: {len(all_rows)}")
    print(f"DUPLICATES FILTERED: {duplicates_count}")

    subjects = {}
    for r in all_rows:
        subjects[r["subject"]] = subjects.get(r["subject"], 0) + 1
    print(f"TOTAL DISTINCT SUBJECTS: {len(subjects)}")

    # Write Master JSON
    json_path = os.path.join(OUTPUT_DIR, "master_taxonomy.json")
    with open(json_path, "w", encoding="utf-8") as f:
        json.dump(all_rows, f, indent=2, ensure_ascii=False)
    print(f"Saved Master JSON: {json_path} ({os.path.getsize(json_path) / 1024 / 1024:.2f} MB)")

    # Write CSV for Google Sheets Master Taxonomy Sheet
    csv_path = os.path.join(OUTPUT_DIR, "master_taxonomy_for_google_sheets.csv")
    with open(csv_path, "w", encoding="utf-8", newline="") as f:
        writer = csv.writer(f)
        writer.writerow(["Sr.", "Subject", "Topic", "Subtopic", "Used", "Used_At", "Times_Shown", "Last_Shown_At", "Video_URL", "Notes", "Active"])
        for r in all_rows:
            writer.writerow([r["sr"], r["subject"], r["topic"], r["subtopic"], "FALSE", "", 0, "", "", "", "TRUE"])
    print(f"Saved Google Sheets CSV: {csv_path} ({os.path.getsize(csv_path) / 1024 / 1024:.2f} MB)")

    # Write Sample Bundle for Instant Frontend Demonstration
    sample_json_path = os.path.join(OUTPUT_DIR, "sample_taxonomy_bundle.json")
    with open(sample_json_path, "w", encoding="utf-8") as f:
        json.dump({
            "stats": {
                "total_subtopics": len(all_rows),
                "available_subtopics": len(all_rows),
                "used_subtopics": 0,
                "used_percentage": 0,
                "used_today": 0
            },
            "subjects_count": len(subjects),
            "sample_records": all_rows[:2000]
        }, f, indent=2, ensure_ascii=False)
    print(f"Saved Sample Bundle: {sample_json_path}")

if __name__ == "__main__":
    main()

File: scripts/prepare-sheets-import.ts
/**
 * ====================================================================
 * KNOWSIGHTS TAXONOMY AGGREGATOR & EXPORTER
 * Extracts all 56,000+ taxonomy rows from source files, normalizes whitespace,
 * removes duplicate rows, preserves original source_sr, and outputs a ready-to-import CSV.
 * ====================================================================
 */

import * as fs from 'fs';
import * as path from 'path';

interface TaxonomyRow {
  sr: number;
  subject: string;
  topic: string;
  subtopic: string;
}

const DOWNLOADS_DIR = 'C:\\Users\\ETIVE\\Downloads';
const OUTPUT_DIR = path.resolve(__dirname, '../data');

if (!fs.existsSync(OUTPUT_DIR)) {
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });
}

function normalize(text: string): string {
  if (!text) return '';
  return text.trim().replace(/\s+/g, ' ');
}

function parseCsv(content: string): TaxonomyRow[] {
  const lines = content.split(/\r?\n/);
  if (lines.length < 2) return [];

  const rows: TaxonomyRow[] = [];
  const headerLine = lines[0].replace(/^\uFEFF/, '').trim();
  const headers = headerLine.split(',').map(h => h.trim().replace(/^["']|["']$/g, ''));

  const srIdx = headers.findIndex(h => /^(sr|sr\.?|serial|id)$/i.test(h));
  const subIdx = headers.findIndex(h => /^subject$/i.test(h));
  const topIdx = headers.findIndex(h => /^topic$/i.test(h));
  const subtopIdx = headers.findIndex(h => /^(subtopic|sub_topic|sub topic)$/i.test(h));

  for (let i = 1; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line) continue;

    // Simple CSV parser supporting quotes
    const cells: string[] = [];
    let inQuotes = false;
    let currentCell = '';

    for (let c = 0; c < line.length; c++) {
      const char = line[c];
      if (char === '"' || char === "'") {
        inQuotes = !inQuotes;
      } else if (char === ',' && !inQuotes) {
        cells.push(currentCell.trim());
        currentCell = '';
      } else {
        currentCell += char;
      }
    }
    cells.push(currentCell.trim());

    const subject = normalize(cells[subIdx !== -1 ? subIdx : 1] || '');
    const topic = normalize(cells[topIdx !== -1 ? topIdx : 2] || '');
    const subtopic = normalize(cells[subtopIdx !== -1 ? subtopIdx : 3] || '');
    const sr = parseInt(cells[srIdx !== -1 ? srIdx : 0], 10) || (rows.length + 1);

    if (subject && topic && subtopic) {
      rows.push({ sr, subject, topic, subtopic });
    }
  }

  return rows;
}

function parseMarkdownTable(content: string, defaultSubject: string): TaxonomyRow[] {
  const lines = content.split(/\r?\n/);
  const rows: TaxonomyRow[] = [];

  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed.startsWith('|') || trimmed.includes('---') || /\|\s*Sr\.\s*\|/i.test(trimmed)) {
      continue;
    }
    const cols = trimmed.split('|').map(c => normalize(c)).filter(c => c.length > 0);
    if (cols.length >= 3) {
      const sr = parseInt(cols[0], 10) || (rows.length + 1);
      let subject = defaultSubject;
      let topic = '';
      let subtopic = '';

      if (cols.length >= 4) {
        subject = cols[1] || defaultSubject;
        topic = cols[2];
        subtopic = cols[3];
      } else {
        topic = cols[1];
        subtopic = cols[2];
      }

      if (topic && subtopic) {
        rows.push({ sr, subject, topic, subtopic });
      }
    }
  }
  return rows;
}

function aggregateAllTaxonomies(): void {
  console.log('🔍 Scanning taxonomy source files in:', DOWNLOADS_DIR);

  const files = fs.readdirSync(DOWNLOADS_DIR);
  const taxonomyFiles = files.filter(f => f.includes('taxonomy') && (f.endsWith('.csv') || f.endsWith('.md')));

  const allRows: TaxonomyRow[] = [];
  const seenKeys = new Set<string>();
  let duplicatesCount = 0;

  for (const f of taxonomyFiles) {
    const fullPath = path.join(DOWNLOADS_DIR, f);
    const content = fs.readFileSync(fullPath, 'utf8');
    let parsed: TaxonomyRow[] = [];

    if (f.endsWith('.csv')) {
      parsed = parseCsv(content);
    } else if (f.endsWith('.md')) {
      const subjectGuess = f.includes('physics') ? 'Physics' : f.includes('economics') ? 'Economics' : 'General Knowledge';
      parsed = parseMarkdownTable(content, subjectGuess);
    }

    let fileAdded = 0;
    for (const row of parsed) {
      const dedupeKey = `${row.subject.toLowerCase()}|${row.topic.toLowerCase()}|${row.subtopic.toLowerCase()}`;
      if (seenKeys.has(dedupeKey)) {
        duplicatesCount++;
        continue;
      }
      seenKeys.add(dedupeKey);
      allRows.push(row);
      fileAdded++;
    }

    console.log(`  ✓ ${f.padEnd(65)} : ${parsed.length} found -> ${fileAdded} unique added`);
  }

  // Re-index sequential source_sr
  allRows.forEach((r, idx) => {
    r.sr = idx + 1;
  });

  console.log('================================================================');
  console.log(`📊 TOTAL UNIQUE TAXONOMY ROWS AGGREGATED: ${allRows.length}`);
  console.log(`🚫 DUPLICATE ROWS FILTERED: ${duplicatesCount}`);
  
  const subjectsMap = new Map<string, number>();
  allRows.forEach(r => {
    subjectsMap.set(r.subject, (subjectsMap.get(r.subject) || 0) + 1);
  });
  console.log(`📚 TOTAL DISTINCT SUBJECTS: ${subjectsMap.size}`);

  // Write Master JSON
  const jsonPath = path.join(OUTPUT_DIR, 'master_taxonomy.json');
  fs.writeFileSync(jsonPath, JSON.stringify(allRows, null, 2), 'utf8');
  console.log(`💾 Saved Master JSON: ${jsonPath} (${(fs.statSync(jsonPath).size / 1024 / 1024).toFixed(2)} MB)`);

  // Write Ready-to-Import CSV for Google Sheets Master Taxonomy tab
  // Headers: Sr.,Subject,Topic,Subtopic,Used,Used_At,Times_Shown,Last_Shown_At,Video_URL,Notes,Active
  const csvPath = path.join(OUTPUT_DIR, 'master_taxonomy_for_google_sheets.csv');
  const csvHeader = 'Sr.,Subject,Topic,Subtopic,Used,Used_At,Times_Shown,Last_Shown_At,Video_URL,Notes,Active\n';
  const csvRows = allRows.map(r => {
    const safeSubj = `"${r.subject.replace(/"/g, '""')}"`;
    const safeTop = `"${r.topic.replace(/"/g, '""')}"`;
    const safeSubtop = `"${r.subtopic.replace(/"/g, '""')}"`;
    return `${r.sr},${safeSubj},${safeTop},${safeSubtop},FALSE,,,0,,,TRUE`;
  }).join('\n');

  fs.writeFileSync(csvPath, csvHeader + csvRows, 'utf8');
  console.log(`💾 Saved Google Sheets CSV: ${csvPath} (${(fs.statSync(csvPath).size / 1024 / 1024).toFixed(2)} MB)`);

  // Also write a rich sample bundle for instant web app testing
  const sampleJsonPath = path.join(OUTPUT_DIR, 'sample_taxonomy_bundle.json');
  const sampleRows = allRows.slice(0, 1500); // 1,500 diverse items for instant offline test
  fs.writeFileSync(sampleJsonPath, JSON.stringify({
    stats: {
      total_subtopics: allRows.length,
      available_subtopics: allRows.length,
      used_subtopics: 0,
      used_percentage: 0,
      used_today: 0
    },
    subjects_count: subjectsMap.size,
    sample_records: sampleRows
  }, null, 2), 'utf8');
  console.log(`💾 Saved Sample Bundle for Web App: ${sampleJsonPath}`);
}

aggregateAllTaxonomies();
